packages/components/select/src/select.js

```diff
@@ -0,0 +1,328 @@
+			/** INTERNAL Latest input value that was used to filter. */
+			filterValue: { type: String, attribute: false },
+			/** INTERNAL Filter to show only selected. */
+			showSelectedOnly: { type: Boolean, reflect: true }
+		};
+	}
+
+	constructor() {
+		super();
+		this.opened = false;
+		this.multiselect = false;
+		this.dirty = false;
+		this.currentSelection = [];
+		this.selected = [];
+		this.inputValue = '';
+		this.filterValue = '';
+		this.showSelectedOnly = false;
+		this.resizeListener = this.updatePosition.bind(this);
+		this.clickOutsideListener = this.handleOutsideClick.bind(this);
+		this.handleInputDebounced = helpers.debounceEvent(() => {
+			this.filterValue = this.inputValue;
+		}, 300);
+	}
+
+	toggle() {
+		this.opened = !this.opened;
+	}
+
+	open() {
+		this.opened = true;
+	}
+
+	firstUpdated() {
+		window.addEventListener('resize', this.resizeListener);
+		window.addEventListener('mousedown', this.clickOutsideListener);
+	}
+
+	disconnectedCallback() {
+		window.removeEventListener('resize', this.resizeListener);
+		window.removeEventListener('mousedown', this.clickOutsideListener);
+	}
+
+	update(changedProperties) {
+		if (changedProperties.has('selected')) {
+			// initialise a current selection with selected items.
+			this.currentSelection = [...this.selected];
+		}
+		if (changedProperties.has('opened')) {
+			// clean filter values on every open/closed state change.
+			this.inputValue = '';
+			this.filterValue = '';
+			this.showSelectedOnly = false;
+		}
+		super.update(changedProperties);
+	}
+
+	attributeChangedCallback(name, oldVal, newVal) {
+		super.attributeChangedCallback(name, oldVal, newVal);
+		switch (name) {
+			case 'opened':
+				if (newVal === null) {
+					// discard changes by assigning previously selected items to current selection
+					this.currentSelection = [...this.selected];
+					this.dirty = false;
+				}
+				if (newVal !== null) {
+					window.customElements.whenDefined('ts-select').then(() => {
+						this.updatePosition();
+					});
+				}
+				break;
+
+			default:
+			// do nothing
+		}
+	}
+
+	/**
+	 * @private
+	 * @param {Event} event
+	 */
+	handleOutsideClick(event) {
+		if (!this.opened) {
+			return;
+		}
+		if (!('composedPath' in event)) {
+			return;
+		}
+		const paths = event.composedPath();
+		const container = this.shadowRoot.getElementById('container');
+		if (paths.indexOf(container) === -1) {
+			this.opened = false;
+		}
+	}
+
+	/**
+	 * Calculates the size and position of the dropdown part.
+	 * @private
+	 */
+	updatePosition() {
+		if (!this.opened) {
+			return;
+		}
+		requestAnimationFrame(() => {
+			const selectInput = this.shadowRoot.getElementById('selectInput');
+			const dropdown = this.shadowRoot.getElementById('dropdown');
+			const listContainer = this.shadowRoot.getElementById('listContainer');
+			if (!dropdown) {
+				// it could happen when element is just initialised but not rendered yet.
+				return;
```

Comment:
nit, I think it would be better to move the selectInput and listContainer after the `if`.